systems/plants/RigidBodyManipulator.cpp

```diff
@@ -2561,6 +2561,172 @@ void RigidBodyManipulator::HandC(MatrixBase<DerivedG> const & q, MatrixBase<Deri
   }
 }
 
+//computes surface tangent vectors for many normal vectors at once
+void RigidBodyManipulator::surfaceTangents(Map<Matrix3xd> const & normals, std::vector< Map<Matrix3xd> > & tangents)
+{
+  const int numContactPairs = normals.cols();
+
+  for(int curNormal = 0 ; curNormal < numContactPairs; curNormal++)
+  {
+    Matrix3kd d;
+    surfaceTangentsSingle(normals.col(curNormal), d);
+    for(int k = 0 ; k < BASIS_VECTOR_HALF_COUNT ; k++)
+    {
+      tangents[k].col(curNormal) = d.col(k);
+    }
+  }
+}
+
+//computes surface tangent vectors for a single normal vector
+void RigidBodyManipulator::surfaceTangentsSingle(Vector3d const & normal, Matrix3kd & d)
+{
+  Vector3d t1,t2;
+  double theta;
+
+  if (1 - normal(2) < EPSILON) { // handle the unit-normal case (since it's unit length, just check z)
+    t1 << 1,0,0;
+  } else if(1 + normal(2) < EPSILON) {
+    t1 << -1,0,0;  //same for the reflected case
+  } else {// now the general case
+    t1 << normal(1), -normal(0) , 0;
+    t1 /= sqrt(normal(1)*normal(1) + normal(0)*normal(0));
+  }
+
+  t2 = t1.cross(normal);
+
+  for (int k=0; k<BASIS_VECTOR_HALF_COUNT; k++) {
+    theta = k*M_PI/BASIS_VECTOR_HALF_COUNT;
+    d.col(k)=cos(theta)*t1 + sin(theta)*t2;
+  }
+}
+
+void RigidBodyManipulator::getUniqueBodiesSorted(VectorXi const & idxA, VectorXi const & idxB, std::vector<int> & bodyIndsSorted)
+{
+  size_t m = idxA.size();
+  std::set<int> bodyInds;
+
+  for(int i = 0 ; i < m ; i++)
+  {  
+    bodyInds.insert(idxA[i]);
+    bodyInds.insert(idxB[i]);
+  }
+
+  bodyIndsSorted.clear();
+
+  for(std::set<int>::const_iterator citer = bodyInds.begin() ; citer != bodyInds.end() ; citer++)
+  {
+    if( *citer > 1 )
+    {
+      bodyIndsSorted.push_back(*citer);
+    }
+  }
+
+  sort(bodyIndsSorted.begin(), bodyIndsSorted.end());
+}
+
+void RigidBodyManipulator::findContactIndexes(VectorXi const & idxList, const int bodyIdx, std::vector<int> & contactIdx)
+{
+  int m = idxList.size();
+  contactIdx.clear();
+  for(int i = 0 ; i < m ; i++) {
+    if(idxList[i] == bodyIdx) {
+      contactIdx.push_back(i); //zero-based index 
+    }
+  }
+}
+
+void RigidBodyManipulator::getBodyPoints(std::vector<int> const & cindA, std::vector<int> const & cindB, Matrix3xd const & xA, Matrix3xd const & xB, MatrixXd & bodyPoints)
+{
+  int i = 0;
+  int numPtsA = cindA.size();
+  int numPtsB = cindB.size();
+
+  bodyPoints.resize(4, numPtsA + numPtsB);
+
+  for(i = 0 ; i < numPtsA ; i++ ) {
+    bodyPoints.col(i) << xA.col(cindA[i]) , 1; //homogeneous coordinates
```

Comment:
For homogeneous coordinates, you can use the .homogeneous() function in MatrixBase, which has an implicit 1 at the end of the vector/matrix while not actually storing it in memory. Probably don't fix in this pull request but keep in mind for later.